src/BSeries.jl

```diff
@@ -1991,8 +2008,19 @@ function _is_energy_preserving_sparse(trees, coefficients)
         # If so, we can exit early since the coefficients of bushy trees
         # must be zero in the modified equation of energy-preserving methods.
         if length(t) > 1 && is_bushy(t)
-            if !iszero(coefficients[t_index])
-                return false
+            if (typeof(coefficients[t_index]) == Float64 || typeof(coefficients[t_index]) == Float32 || typeof(coefficients[t_index]) == Float16) && tol == 0
+                tol = 1e-14
+                if !isapprox(coefficients[t_index], 0, atol = tol)
+                    return false
+                end
+            elseif tol != 0
+                if !isapprox(coefficients[t_index], 0, atol = tol)
+                    return false
+                end
+            else
+                if !iszero(coefficients[t_index])
+                    return false
+                end
```

Comment:
I do not like that this code is duplicated. You could move it to a small function and call it from both `_is_energy_preserving_sparse` and `_is_energy_preserving_dense`.
Moreover, I think the way how you handle the default tolerance can be improved. 
- First, if the user selects `tol = 0`, we should respect that and use `tol = 0` - as described in the docstring.
- Next, there are much more floating point types - you could check `<: AbstractFloat` instead of `== Float64` etc.
- I gave you a concrete suggestion how to handle this in the previous PR:
  ```julia
  V = eltype(series_integrator)
  if tol === nothing
    if V <: AbstractFloat
      tol = 100 * eps(T)
    else
      tol = zero(T)
    end
  end
  ```
  _Originally posted in https://github.com/ranocha/BSeries.jl/pull/156#pullrequestreview-1609538103_
  Is there something wrong with such an approach in your opinion? To adapt it to this PR, you could create a new function, say `energy_preserving_default_tolerance(V, tol)`, that contains the logic above given the `valtype(series_integrator)` etc. This has to be documented appropriately in the docstrings of the high-level functions. Then, these high-level functions would have the default value `tol = nothing` that is converted appropriately by the helper function `energy_preserving_default_tolerance`.

Reply:
``````Okey, I can implement this function. Do you want to use the function `isapprox()` always then? I discussed with David about using `iszero()` for non - `AbstractFloat` inputs, but if we set a tolerance using this function
`function energy_preserving_default_tolerance(V, tol)
    if tol == 0
        if V <: AbstractFloat
            tol = 100 * eps(T)
        else
            tol = zero(T)
        end
    end
    return tol
end`
then we can use only the function `isapprox()` and only once.

Reply:
Yes, that's fine with me. It should simplify the control flow.

Reply:
Okey. Please take a look at the new changes I just pushed and tell me what you think.